### PR TITLE
Backport 11167 (partial), 11372, 11630 and 14242

### PR DESCRIPTION
**Pull request by UdjinM6**
Individual commits are from https://github.com/bitcoin/bitcoin/pull/11167, should be easier to review it this way imo cause half of original commits was omitted (due to segwit-only code). Can squash into one "Partial merge 11167..." after reviews.

### CODE REVIEW

**Review by xdustinface — APPROVED** [review]
Looks good, utACK.

Any plans in your mind about actually using bech32?

File: src/key_io.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+    std::string operator()(const CNoDestination& no) const { return {}; }
+};
+
+CTxDestination DecodeDestination(const std::string& str, const CChainParams& params)
+{
+    std::vector<unsigned char> data;
+    uint160 hash;
+    if (DecodeBase58Check(str, data)) {
+        // base58-encoded Bitcoin addresses.
```
Comment (PastaPastaPasta) [inline src/key_io.cpp:51]: dashify?

**Review by PastaPastaPasta — COMMENTED** [review]
Generally looks good, one comment, also curious about dustin's question

**Comment by thephez** [comment]
> Any plans in your mind about actually using bech32?

I was wondering about this and other alternatives too (e.g. the Bitcoin Cash format which is based in part on bech32 work).

**Comment by UdjinM6** [comment]
> > Any plans in your mind about actually using bech32?
> 
> I was wondering about this and other alternatives too (e.g. the Bitcoin Cash format which is based in part on bech32 work).

I can't think of any reason to go that route atm tbh. If you have any ideas about how this would be helpful (keeping in mind the disruption it's going to cause), I'll gladly reconsider it :)

**Review by xdustinface — DISMISSED** [review]
utACK

> I can't think of any reason to go that route atm tbh. If you have any ideas about how this would be helpful (keeping in mind the disruption it's going to cause), I'll gladly reconsider it :)

I just don't think there will ever be a "good/better time" to go that route and work out/handle the disruptions it brings. IMO its not a bad thing to introduce a new address type based on bech32 alongside with our current format just because it's an improved format for some (more or less significant) reasons and improving things always makes sense :) I mean, its not that it suddenly makes usernames out of addresses but still better than what we currenly have. There is no pressure to force it to be used in the near future since they can just co-exists, we can still default to the old format for the meantime and providing tools to convert between the formats is also not a big deal in RPC, GUI or Web (The disruptions will probably mostly come up for the support? 🙈). I also likely miss some of the issues you have in mind so let me know your concerns if so.

At the end, i won't fight very hard for it but i would definitely vote for it.

**Comment by PastaPastaPasta** [comment]
Looks good, please squash

**Comment by UdjinM6** [comment]
Squashed

**Review by xdustinface — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK, squash clean

**Pull request merged**